thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/cache/DefaultTimeSeriesCache.java

```diff
@@ -0,0 +1,194 @@
+ */
+
+public class DefaultTimeSeriesCache implements TimeSeriesCache {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultTimeSeriesCache.class);
+
+  private final MetricConfigManager metricDAO;
+  private final DatasetConfigManager datasetDAO;
+  private final QueryCache queryCache;
+  private CouchbaseCacheDAO cacheDAO;
+  private final ExecutorService executor;
+
+  public DefaultTimeSeriesCache(MetricConfigManager metricDAO, DatasetConfigManager datasetDAO, QueryCache queryCache, CouchbaseCacheDAO cacheDAO, ExecutorService executorService) {
+    this.metricDAO = metricDAO;
+    this.datasetDAO = datasetDAO;
+    this.queryCache = queryCache;
+    this.cacheDAO = cacheDAO;
+    this.executor = executorService;
+  }
+
+  /**
+   * Given a ThirdEyeRequest object, builds the CacheRequest object and attempts to fetch data
+   * from the cache. If the requested slice of data is not in the cache or is only partially in
+   * the cache, will fetch the missing slices and build the complete timeseries.
+   * @param thirdEyeRequest ThirdEyeRequest built from aligned metricSlice request.
+   * @return ThirdEyeResponse, the full response for the request
+   * @throws Exception if fetch from original data source was not successful.
+   */
+  public ThirdEyeResponse fetchTimeSeries(ThirdEyeRequest thirdEyeRequest) throws Exception {
+
+    if (!CacheConfig.useCentralizedCache()) {
+      return this.queryCache.getQueryResult(thirdEyeRequest);
+    }
+
+    ThirdEyeCacheResponse cacheResponse = cacheDAO.tryFetchExistingTimeSeries(ThirdEyeCacheRequest.from(thirdEyeRequest));
+
+    DateTime sliceStart = thirdEyeRequest.getStartTimeInclusive();
+    DateTime sliceEnd = thirdEyeRequest.getEndTimeExclusive();
+
+    if (cacheResponse.isMissingSlice(sliceStart.getMillis(), sliceEnd.getMillis())) {
+      fetchMissingSlices(cacheResponse);
+    }
+
+    return buildThirdEyeResponseFromCacheResponse(cacheResponse);
+  }
+
+  /**
+   * Used if cache had partial or no data for the requested time slice. Checks for which
+   * slices are missing and fetches them, then adds them to the response.
+   * @param cacheResponse cache response object
+   * @throws Exception if fetch to data source throws error.
+   */
+  private void fetchMissingSlices(ThirdEyeCacheResponse cacheResponse) throws Exception {
+
+    ThirdEyeRequest request = cacheResponse.getCacheRequest().getRequest();
+    long metricId = request.getMetricFunctions().get(0).getMetricId();
+    long requestSliceStart = request.getStartTimeInclusive().getMillis();
+    long requestSliceEnd = request.getEndTimeExclusive().getMillis();
+
+    ThirdEyeResponse result;
+    MetricSlice slice;
+
+    if (cacheResponse.hasNoRows()) {
+      result = this.queryCache.getQueryResult(cacheResponse.getCacheRequest().getRequest());
+      insertTimeSeriesIntoCache(result);
+      cacheResponse.mergeSliceIntoRows(result);
+    }
+
+    if (cacheResponse.isMissingStartSlice(requestSliceStart)) {
```

Comment:
It's possible for the data in the cache to be missing some data from the beginning of the timeseries and also missing some data from the end of the timeseries at the same time, I think? It's a pretty rare case, though, since we need to fetch 28/90 days training data, but can still happen due to TTL/expiration in the cache if older documents don't have expiration refreshed and thus expire.

Reply:
oh, I see your point. I will add this, thanks. Otherwise if we are doing detection from 10/01 - 11/01 but dataset only has data until 10/31, we will try to fetch last days data again even though it doesn't exist.

Reply:
done